**Update version 28/12/2020 following videos of Lesson 4 Section 5 and checking notebook provided**

# Azure Machine Learning Pipeline with AutoMLStep 

This notebook is part of [Project 2: Operationalizing Machine Learning with Azure](https://github.com/dpbac/Operationalizing-Machine-Learning-with-Azure) of the `Udacity Nanodegree Program : Machine Learning Engineer with Azure`.

In this project you will continue working with the Bank Dataset introduced in [Project 1: Optimizing an ML Pipeline in Azure](https://github.com/dpbac/Optimizing-an-ML-Pipeline-in-Azure). We will use Azure in this project to configure a cloud-based machine learning product model, deploy it, and consume it.

**REVIEW AND COMPLETE ALL TEXT BELOW**

This notebook demonstrates the use of AutoMLStep in Azure Machine Learning Pipeline.

A model is generated using AutoML for classifcation using the dataset available at https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv

## Introduction

In this example we showcase how you can use AzureML Dataset to load data for AutoML via AML Pipeline. 

If you are using an Azure Machine Learning Notebook VM, you are all set. Otherwise, make sure you have executed the [configuration](https://aka.ms/pl-config) before running this notebook.

In this notebook the following steps are performed

1. Create an `Experiment` in an existing `Workspace`.
2. Create or Attach existing AmlCompute to a workspace.
3. Define data loading in a `TabularDataset`.
4. Configure AutoML using `AutoMLConfig`.
5. Use AutoMLStep
6. Train the model using AmlCompute
7. Explore the results.
8. Test the best fitted model.



## Azure Machine Learning and Pipeline SDK-specific imports

In [ ]:
import logging
import os
import csv

from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
from sklearn import datasets
import pkg_resources

import azureml.core
from azureml.core.experiment import Experiment
from azureml.core.workspace import Workspace
from azureml.train.automl import AutoMLConfig
from azureml.core.dataset import Dataset

from azureml.pipeline.steps import AutoMLStep

# Check core SDK version number
print("SDK version:", azureml.core.VERSION)

In [1]:
# # Azure Machine Learning and Pipeline SDK-specific imports

# import logging
# import os
# import csv

# from matplotlib import pyplot as plt
# import numpy as np
# import pandas as pd
# from sklearn import datasets
# import pkg_resources

# import azureml.core
# from azureml.core.experiment import Experiment
# from azureml.core.workspace import Workspace
# from azureml.train.automl import AutoMLConfig
# from azureml.core.dataset import Dataset

# from azureml.pipeline.steps import AutoMLStep

# # Check core SDK version number
# print("SDK version:", azureml.core.VERSION)

SDK version: 1.19.0


## Initialize Workspace
Initialize a workspace object from persisted configuration. Make sure the config file is present at .\config.json

The config.json can be downloaded in the overview of Azure portal. (**Do I need to download?**)

In [2]:
ws = Workspace.from_config()
print(ws.name, ws.resource_group, ws.location, ws.subscription_id, sep = '\n')

quick-starts-ws-132494
aml-quickstarts-132494
southcentralus
9e65f93e-bdd8-437b-b1e8-0647cd6098f7


## Create an Azure ML experiment

**REVIEW AND UPDATE TEXT**
**ATTENTION NAME OF THE EXPERIMENT**

Let's create an experiment named "automlstep-classification" and a folder to hold the training scripts. The script runs will be recorded under the experiment in Azure.

The best practice is to use separate folders for scripts and its dependent files for each step and specify that folder as the `source_directory` for the step. This helps reduce the size of the snapshot created for the step (only the specific folder is snapshotted). Since changes in any files in the `source_directory` would trigger a re-upload of the snapshot, this helps keep the reuse of the step when there are no changes in the `source_directory` of the step.

*Udacity Note:* There is no need to create an Azure ML experiment, this needs to re-use the experiment that was already created

In [3]:
# Choose a name for the run history container in the workspace.
# NOTE: update these to match your existing exp name
experiment_name = 'automl-experiment'
project_folder = './pipeline-project'

experiment = Experiment(ws, experiment_name)

experiment

Experiment(Name: automl-experiment,
Workspace: quick-starts-ws-132494)

In [4]:
dic_data = {'Workspace name': ws.name,
            'Azure region': ws.location,
            'Subscription id': ws.subscription_id,
            'Resource group': ws.resource_group,
            'Experiment Name': experiment.name}

df_data = pd.DataFrame.from_dict(data = dic_data, orient='index')

df_data.rename(columns={0:''}, inplace = True)
df_data

Workspace name                 quick-starts-ws-132494
Azure region                           southcentralus
Subscription id  9e65f93e-bdd8-437b-b1e8-0647cd6098f7
Resource group                 aml-quickstarts-132494
Experiment Name                     automl-experiment

### Create or Attach an AmlCompute cluster

Now that we have initialized our workspace and created our experiment, it is time to define our resources. This means we need to create a [compute target](https://docs.microsoft.com/azure/machine-learning/service/concept-azure-machine-learning-architecture#compute-target) for your AutoML run.

In this section you will create default compute clusters for use by the notebook and any other necessary operations we need. Here we get the default `AmlCompute` as your training compute resource.

In order to create a cluster we need to specify a compute configuration that defines the `type of machine` to be used and the `scalability behaviors`. Also, it is necessary to define the name of the cluster which must be unique within the workspace. This name is used to address the cluster later.

For this project we use a CPU cluster with following parameters:

* `type of the machine`:

    * `vm_size`: Defines the size of the virtual machine. We use here "STANDARD_DS12_V2" (more details [here](https://docs.microsoft.com/en-us/azure/cloud-services/cloud-services-sizes-specs#dv2-series))

* `Scalability behaviors`:

    * `min_nodes`: Sets minimun size of the cluster. Setting the minimum to 0 the cluster will shut down all nodes while not in use. If you use another value you are able to have faster start-up times, but you will also be billed when the cluster is not in use. In this experiment we define `min_nodes` as 1.

    * `max_nodes`: Sets the maximun size of the cluster. Larger number allows for more concurrency and a greater distributed processing of scale-out jobs.

In [8]:
from azureml.core.compute import ComputeTarget, AmlCompute
from azureml.core.compute_target import ComputeTargetException

# Define CPU cluster name
cpu_cluster_name = "cpu-cluster"


# Verify that cluster does not exist already
try:
    cpu_cluster = ComputeTarget(workspace=ws, name=cpu_cluster_name)
    print("Found existing cpu-cluster. Use it.")
except ComputeTargetException:
    # Specify the configuration for the new cluster
    compute_config = AmlCompute.provisioning_configuration(vm_size="STANDARD_DS12_V2",
                                                           #vm_priority = 'lowpriority', # optional
                                                           min_nodes=1, # when innactive
                                                           max_nodes=4) # when busy

    # Create the cluster with the specified name and configuration
    cpu_cluster = ComputeTarget.create(ws, cpu_cluster_name, compute_config)

# can poll for a minimum number of nodes and for a specific timeout. 
# if no min node count is provided it will use the scale settings for the cluster

# cpu_cluster.wait_for_completion(show_output=True, min_node_count = 1, timeout_in_minutes = 10)
cpu_cluster.wait_for_completion(show_output=True)

# For a more detailed view of current AmlCompute status, use get_status()
print(cpu_cluster.get_status().serialize())

Found existing cpu-cluster. Use it.

Running
{'errors': [], 'creationTime': '2020-12-30T14:51:32.252246+00:00', 'createdBy': {'userObjectId': '0f08476b-7bef-4402-9c0d-39ea8bdaae40', 'userTenantId': '660b3398-b80e-49d2-bc5b-ac1dc93b5254', 'userName': None}, 'modifiedTime': '2020-12-30T14:54:35.027894+00:00', 'state': 'Running', 'vmSize': 'STANDARD_DS12_V2'}


In [9]:
# Check details about compute_targets (i.e. cpu_cluster)

compute_targets = ws.compute_targets
for name, ct in compute_targets.items():
    print(name, ct.type, ct.provisioning_state)

cpu-cluster ComputeInstance Succeeded


## Dataset

The code below try first to load the dataset from the workspace. The key is the dataset name and is used to find the dataset and load it.

If the key is not found, i.e., the dataset is not in the workspace the dataset is loaded from the web using the link given at the beginning of this notebook.

In [10]:
# Try to load the dataset from the Workspace. Otherwise, create it from the file
# NOTE: update the key to match the dataset name
found = False
key = "BankMarketing"
description_text = "Bank Marketing DataSet for 2nd Project Udacity Azure ML Engineer."

if key in ws.datasets.keys(): 
        found = True
        dataset = ws.datasets[key] 

if not found:
        # Create AML Dataset and register it into Workspace
        example_data = 'https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv'
        # Create TabularDataset using TabularDatasetFactory
        dataset = Dataset.Tabular.from_delimited_files(example_data)        
        #Register Dataset in Workspace
        dataset = dataset.register(workspace=ws,
                                   name=key,
                                   description=description_text)


#create a data frame
df = dataset.to_pandas_dataframe()

### Inspect Dataset

In [11]:
df.head()

age          job  marital    education  default housing loan    contact  \
0   57   technician  married  high.school       no      no  yes   cellular   
1   55      unknown  married      unknown  unknown     yes   no  telephone   
2   33  blue-collar  married     basic.9y       no      no   no   cellular   
3   36       admin.  married  high.school       no      no   no  telephone   
4   27    housemaid  married  high.school       no     yes   no   cellular   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         1      failure         -1.8   
1   may         thu  ...         2    999         0  nonexistent          1.1   
2   may         fri  ...         1    999         1      failure         -1.8   
3   jun         fri  ...         4    999         0  nonexistent          1.4   
4   jul         fri  ...         2    999         0  nonexistent          1.4   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          92.893          -46.2      1.299       5099.1  no  
1          93.994          -36.4      4.860       5191.0  no  
2          92.893          -46.2      1.313       5099.1  no  
3          94.465          -41.8      4.967       5228.1  no  
4          93.918          -42.7      4.963       5228.1  no  

[5 rows x 21 columns]

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32950 entries, 0 to 32949
Data columns (total 21 columns):
age               32950 non-null int64
job               32950 non-null object
marital           32950 non-null object
education         32950 non-null object
default           32950 non-null object
housing           32950 non-null object
loan              32950 non-null object
contact           32950 non-null object
month             32950 non-null object
day_of_week       32950 non-null object
duration          32950 non-null int64
campaign          32950 non-null int64
pdays             32950 non-null int64
previous          32950 non-null int64
poutcome          32950 non-null object
emp.var.rate      32950 non-null float64
cons.price.idx    32950 non-null float64
cons.conf.idx     32950 non-null float64
euribor3m         32950 non-null float64
nr.employed       32950 non-null float64
y                 32950 non-null object
dtypes: float64(5), int64(5), object(11)
memory usa

In [13]:
df.describe()

age      duration      campaign         pdays      previous  \
count  32950.000000  32950.000000  32950.000000  32950.000000  32950.000000   
mean      40.040212    257.335205      2.561730    962.174780      0.174780   
std       10.432313    257.331700      2.763646    187.646785      0.496503   
min       17.000000      0.000000      1.000000      0.000000      0.000000   
25%       32.000000    102.000000      1.000000    999.000000      0.000000   
50%       38.000000    179.000000      2.000000    999.000000      0.000000   
75%       47.000000    318.000000      3.000000    999.000000      0.000000   
max       98.000000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  32950.000000    32950.000000   32950.000000  32950.000000  32950.000000  
mean       0.076228       93.574243     -40.518680      3.615654   5166.859608  
std        1.572242        0.578636       4.623004      1.735748     72.208448  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

In [14]:
df.y.value_counts(normalize=True)

no     0.887951
yes    0.112049
Name: y, dtype: float64

Unbalanced data: 88.80% label 'no' and 11.20% label 'yes'.

We can apply some techniques on the dataset to make the data more balanced before applying ML. **Automated ML** has built in capabilities to help deal with imbalanced. For instances, the algorithms used by automated ML detect imbalance when the number of samples in the minority class is equal to or fewer than 20% of the number of samples in the majority class, where minority class refers to the one with fewest samples and majority class refers to the one with most samples. Subsequently, AutoML will run an experiment with sub-sampled data to check if using class weights would remedy this problem and improve performance.

In addition, it is important to use a more appropriated metric such as **AUC_weighted** which is a primary metric that calculates the contribution of every class based on the relative number of samples representing that class. Therefore, it is more robust against imbalance.

For more details, check this [link](https://docs.microsoft.com/en-us/azure/machine-learning/concept-manage-ml-pitfalls).



## Train model:

Now we use the [BankMarketing Dataset](https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv) to obtain a model by running AutoML.

AutoML is an automated machine learning capability included in Azure ML. It leverages the scalability of cloud compute to automatically try multiple pre-processing techniques and model-training algorithms in parallel to find the best performing supervised machine learning model for a dataset. In this particular case for the `BankMarketing` dataset, we have just loaded.

### Configure AutoML

In [15]:
automl_settings = {
    "experiment_timeout_minutes": 60, # define how long, in minutes, the experiment should continue to run.
    "max_concurrent_iterations": 5,
    "primary_metric" : 'AUC_weighted'
}
automl_config = AutoMLConfig(compute_target=cpu_cluster,
                             task = "classification",
                             training_data=dataset,
                             label_column_name="y",   
                             path = project_folder,
                             enable_early_stopping= True,
                             featurization= 'auto', # as part of preprocessing, data guardrails and featurization steps are performed automatically.
                             debug_log = "automl_errors.log",
                             **automl_settings
                            )

### Create Pipeline and AutoMLStep

You can define outputs for the AutoMLStep using TrainingOutput.

In [16]:
from azureml.pipeline.core import PipelineData, TrainingOutput

ds = ws.get_default_datastore()
metrics_output_name = 'metrics_output'
best_model_output_name = 'best_model_output'

metrics_data = PipelineData(name='metrics_data',
                           datastore=ds,
                           pipeline_output_name=metrics_output_name,
                           training_output=TrainingOutput(type='Metrics'))
model_data = PipelineData(name='model_data',
                           datastore=ds,
                           pipeline_output_name=best_model_output_name,
                           training_output=TrainingOutput(type='Model'))

Create an AutoMLStep.

In [17]:
automl_step = AutoMLStep(
    name='automl_module',
    automl_config=automl_config,
    outputs=[metrics_data, model_data],
    allow_reuse=True)

Now we create the pipeline using the `automl_step` just created.

In [18]:
from azureml.pipeline.core import Pipeline
pipeline = Pipeline(
    description="pipeline_with_automlstep",
    workspace=ws,    
    steps=[automl_step])

In [19]:
pipeline_run = experiment.submit(pipeline)

Created step automl_module [4dbc014a][78d48988-832b-4531-ae76-89e8edf0022a], (This step will run and generate new outputs)
Submitted PipelineRun 92546f87-156f-4e2f-b526-db3b8b637433
Link to Azure Machine Learning Portal: https://ml.azure.com/experiments/automl-experiment/runs/92546f87-156f-4e2f-b526-db3b8b637433?wsid=/subscriptions/9e65f93e-bdd8-437b-b1e8-0647cd6098f7/resourcegroups/aml-quickstarts-132494/workspaces/quick-starts-ws-132494


In [20]:
from azureml.widgets import RunDetails
RunDetails(pipeline_run).show()

_PipelineWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', …

In [21]:
pipeline_run.wait_for_completion()

This usually indicates a package conflict with one of the dependencies of azureml-core or azureml-pipeline-core.
Please check for package conflicts in your python environment


PipelineRunId: 92546f87-156f-4e2f-b526-db3b8b637433
Link to Azure Machine Learning Portal: https://ml.azure.com/experiments/automl-experiment/runs/92546f87-156f-4e2f-b526-db3b8b637433?wsid=/subscriptions/9e65f93e-bdd8-437b-b1e8-0647cd6098f7/resourcegroups/aml-quickstarts-132494/workspaces/quick-starts-ws-132494
PipelineRun Status: Running




PipelineRun Execution Summary
PipelineRun Status: Finished
{'runId': '92546f87-156f-4e2f-b526-db3b8b637433', 'status': 'Completed', 'startTimeUtc': '2020-12-30T14:59:31.967592Z', 'endTimeUtc': '2020-12-30T15:31:26.605159Z', 'properties': {'azureml.runsource': 'azureml.PipelineRun', 'runSource': 'SDK', 'runType': 'SDK', 'azureml.parameters': '{}'}, 'inputDatasets': [], 'outputDatasets': [], 'logFiles': {'logs/azureml/executionlogs.txt': 'https://mlstrg132494.blob.core.windows.net/azureml/ExperimentRun/dcid.92546f87-156f-4e2f-b526-db3b8b637433/logs/azureml/executionlogs.txt?sv=2019-02-02&sr=b&sig=UcVlbT6dgvTaFeRpZfsT2mT%2Bg65qYo6ku7yoLpKiFKo%3D&st=2

'Finished'

In [ ]:
RunDetails(pipeline_run).show()

## Examine Results

### Retrieve the metrics of all child runs
Outputs of above run can be used as inputs of other steps in pipeline. 

In this section we test the pipeline and retrieve the best model.

In [22]:
metrics_output = pipeline_run.get_pipeline_output(metrics_output_name)
num_file_downloaded = metrics_output.download('.', show_progress=True)

Downloaded azureml/f1c74be9-d36e-4365-8d50-0248a25c6dd7/metrics_data, 1 files out of an estimated total of 1


In [23]:
import json
with open(metrics_output._path_on_datastore) as f:
    metrics_output_result = f.read()
    
deserialized_metrics_output = json.loads(metrics_output_result)
df = pd.DataFrame(deserialized_metrics_output)
df

f1c74be9-d36e-4365-8d50-0248a25c6dd7_0  \
balanced_accuracy                                  [0.7513392683482543]   
norm_macro_recall                                  [0.5026785366965085]   
average_precision_score_macro                      [0.8151093723721079]   
AUC_weighted                                       [0.9450464668693167]   
recall_score_macro                                 [0.7513392683482543]   
weighted_accuracy                                  [0.9514937218005303]   
f1_score_micro                                     [0.9116843702579667]   
recall_score_weighted                              [0.9116843702579667]   
AUC_macro                                          [0.9450464668693166]   
precision_score_micro                              [0.9116843702579667]   
average_precision_score_micro                      [0.9806603102489483]   
precision_score_macro                              [0.7819118765348991]   
log_loss                                          [0.17775706110025447]   
f1_score_weighted                                  [0.9091539479147899]   
accuracy                                           [0.9116843702579667]   
recall_score_micro                                 [0.9116843702579667]   
precision_score_weighted                           [0.9072720074188747]   
matthews_correlation                               [0.5323740218566827]   
f1_score_macro                                     [0.7653697272147331]   
AUC_micro                                           [0.979695082216353]   
average_precision_score_weighted                   [0.9531771295804466]   

                                 f1c74be9-d36e-4365-8d50-0248a25c6dd7_2  \
balanced_accuracy                                  [0.5509658292071642]   
norm_macro_recall                                 [0.10193165841432839]   
average_precision_score_macro                      [0.7528045021262322]   
AUC_weighted                                       [0.9020014003967792]   
recall_score_macro                                 [0.5509658292071642]   
weighted_accuracy                                  [0.9822991003188568]   
f1_score_micro                                     [0.8965098634294385]   
recall_score_weighted                              [0.8965098634294385]   
AUC_macro                                          [0.9020014003967791]   
precision_score_micro                              [0.8965098634294385]   
average_precision_score_micro                      [0.9692353653238829]   
precision_score_macro                              [0.8391525423728814]   
log_loss                                          [0.23826156370538618]   
f1_score_weighted                                  [0.8597893321988491]   
accuracy                                           [0.8965098634294385]   
recall_score_micro                                 [0.8965098634294385]   
precision_score_weighted                           [0.8850563514312904]   
matthews_correlation                              [0.26294631048753514]   
f1_score_macro                                     [0.5654495254382943]   
AUC_micro                                          [0.9681995758506589]   
average_precision_score_weighted                   [0.9339787291727356]   

                                 f1c74be9-d36e-4365-8d50-0248a25c6dd7_23  \
balanced_accuracy                                   [0.7459312545962097]   
norm_macro_recall                                  [0.49186250919241936]   
average_precision_score_macro                       [0.8130064306176901]   
AUC_weighted                                        [0.9423114326837048]   
recall_score_macro                                  [0.7459312545962097]   
weighted_accuracy                                   [0.9539729009155015]   
f1_score_micro                                      [0.9125948406676783]   
recall_score_weighted                               [0.9125948406676783]   
AUC_macro                                      

### Retrieve the Best Model

In [24]:
# Retrieve best model from Pipeline Run
best_model_output = pipeline_run.get_pipeline_output(best_model_output_name)
num_file_downloaded = best_model_output.download('.', show_progress=True)

Downloaded azureml/f1c74be9-d36e-4365-8d50-0248a25c6dd7/model_data, 1 files out of an estimated total of 1


In [25]:
import pickle

with open(best_model_output._path_on_datastore, "rb" ) as f:
    best_model = pickle.load(f)
best_model

PipelineWithYTransformations(Pipeline={'memory': None,
                                       'steps': [('datatransformer',
                                                  DataTransformer(enable_dnn=None,
                                                                  enable_feature_sweeping=None,
                                                                  feature_sweeping_config=None,
                                                                  feature_sweeping_timeout=None,
                                                                  featurization_config=None,
                                                                  force_text_dnn=None,
                                                                  is_cross_validation=None,
                                                                  is_onnx_compatible=None,
                                                                  logger=None,
                                                              

In [26]:
best_model.steps

[('datatransformer',
  DataTransformer(enable_dnn=None, enable_feature_sweeping=None,
                  feature_sweeping_config=None, feature_sweeping_timeout=None,
                  featurization_config=None, force_text_dnn=None,
                  is_cross_validation=None, is_onnx_compatible=None, logger=None,
                  observer=None, task=None, working_dir=None)),
 ('prefittedsoftvotingclassifier',
  PreFittedSoftVotingClassifier(classification_labels=None,
                                estimators=[('27',
                                             Pipeline(memory=None,
                                                      steps=[('standardscalerwrapper',
                                                              <azureml.automl.runtime.shared.model_wrappers.StandardScalerWrapper object at 0x7f1c90be56d8>),
                                                             ('xgboostclassifier',
                                                              XGBoostClassifier(ba

### Test the Model
#### Load Test Data


In [27]:
dataset_test = Dataset.Tabular.from_delimited_files(path='https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv')
df_test = dataset_test.to_pandas_dataframe()
df_test = df_test[pd.notnull(df_test['y'])]

y_test = df_test['y']
X_test = df_test.drop(['y'], axis=1)

#### Testing Our Best Fitted Model

We will use confusion matrix to see how our model works.

In [34]:
from sklearn.metrics import confusion_matrix
ypred = best_model.predict(X_test)
cm = confusion_matrix(y_test, ypred)

In [35]:
# Visualize the confusion matrix
pd.DataFrame(cm).style.background_gradient(cmap='Blues', low=0, high=0.9)

## Publish and run from REST endpoint

Publishing a pipeline is the process of making a pipeline publicly available. Here we will use Python SDK to publish our pipeline.

The following code publish the pipeline to your workspace. In your workspace in the portal, you can see metadata for the pipeline including run history and durations. You can also run the pipeline manually from the portal.

Additionally, publishing a pipeline, a public HTTP endpoint becomes available, allowing other services, including external ones, to interact with an Azure Pipeline.

In [36]:
published_pipeline = pipeline_run.publish_pipeline(
    name="Bankmarketing Train", 
    description="Training bankmarketing pipeline", 
    version="1.0")

published_pipeline

Pipeline(Name: Bankmarketing Train,
Id: b551698f-dae4-4b45-9735-f915353e98b5,
Status: Active,
Endpoint: https://southcentralus.api.azureml.ms/pipelines/v1.0/subscriptions/9e65f93e-bdd8-437b-b1e8-0647cd6098f7/resourceGroups/aml-quickstarts-132494/providers/Microsoft.MachineLearningServices/workspaces/quick-starts-ws-132494/PipelineRuns/PipelineSubmit/b551698f-dae4-4b45-9735-f915353e98b5)

Authenticate once again, to retrieve the `auth_header` so that the endpoint can be used

In [37]:
from azureml.core.authentication import InteractiveLoginAuthentication

interactive_auth = InteractiveLoginAuthentication()
auth_header = interactive_auth.get_authentication_header()

Get the `REST url` from the endpoint property of the published pipeline object. You can also find the REST url in your workspace in the portal. Build an HTTP POST request to the endpoint, specifying your authentication header. Additionally, add a JSON payload object with the experiment name and the batch size parameter. As a reminder, the `process_count_per_node` is passed through to `ParallelRunStep` because you defined it is defined as a `PipelineParameter` object in the step configuration.

Make the request to trigger the run. Access the Id key from the response dict to get the value of the run id.

In [38]:
import requests

rest_endpoint = published_pipeline.endpoint
response = requests.post(rest_endpoint, 
                         headers=auth_header, 
                         json={"ExperimentName": "pipeline-rest-endpoint"}
                        )

In [39]:
try:
    response.raise_for_status()
except Exception:    
    raise Exception("Received bad response from the endpoint: {}\n"
                    "Response Code: {}\n"
                    "Headers: {}\n"
                    "Content: {}".format(rest_endpoint, response.status_code, response.headers, response.content))

run_id = response.json().get('Id')
print('Submitted pipeline run: ', run_id)

Submitted pipeline run:  72e2446f-ef65-4d60-9205-9099d08f9547


Use the run id to monitor the status of the new run. This will take another 10-15 min to run and will look similar to the previous pipeline run, so if you don't need to see another pipeline run, you can skip watching the full output.

In [40]:
from azureml.pipeline.core.run import PipelineRun
from azureml.widgets import RunDetails

published_pipeline_run = PipelineRun(ws.experiments["pipeline-rest-endpoint"], run_id)
RunDetails(published_pipeline_run).show()

_PipelineWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', …